src/System.IO.FileSystem/src/System/IO/FindEnumerable.Windows.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Runtime.ConstrainedExecution;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.IO
+{
+    internal partial class FindEnumerable<T> : IEnumerable<T>
```

Comment:
Any reason this couldn't implement both `IEnumerable<T>` and `IEnumerator<T>`, and return itself from `GetEnumerator` in the common case to avoid the enumerator allocation, like `Win32FileSystemEnumerableIterator<TSource>` before this change?

Reply:
I was thinking of being friendly to Linq queries being run multiple times. I'll think it through some more.

Reply:
It should still support being able to call `GetEnumerator` multiple times... in such cases, you'd create a new instance with initial state.

Reply:
> It should still support being able to call GetEnumerator multiple times... in such cases, you'd create a new instance with initial state.

Not sure I follow. How would that save an allocation?

Reply:
Take a look at [`Iterator<TSource>`](https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System/IO/Iterator.cs) (which is the base class of `Win32FileSystemEnumerableIterator<TSource>`) :

`Iterator<TSource>` implements both `IEnumerable<T>` and `IEnumerator<T>`.

Here's the `GetEnumerator` implementation:

https://github.com/dotnet/corefx/blob/db1793c43b1c681442ec8035eefa7a41e50abffb/src/System.IO.FileSystem/src/System/IO/Iterator.cs#L43-L54

The common case when using `foreach` or a LINQ query is to get the enumerable and then immediately call `GetEnumerator` to get the enumerator to start enumerating through the enumerable (e.g. `foreach (string file in Directory.EnumerateFiles(...)) { ... }`). If `GetEnumerator` hasn't already been called and it's the same thread, it can simply return itself. Otherwise, a new instance is created.

This is also how iterators work in C# when you use `yield return`. Similar iterator code is [emitted](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABADAAmwEYAWAbgFgAoa7AZkKIDZCAmfAYWoG9r9/CDYi2J0APAEsAdjAB8+AGIB7JQAoAlHwG8qAvYyKEA7PiKVd+/sUPYTrc5atEbJug4EBfah6A=) by the compiler to use the same instance for both the enumerable and enumerator, only allocating a new instance if `GetEnumerator` if necessary.

Reply:
Fair enough- I'll look at utilizing `Iterator`.

Reply:
Whether or not `Iterator` is used as a base class, the same approach of reusing the enumerable object as the enumerator object for at least the initial enumeration should definitely be done.

Reply:
Making the change now.